xla/service/gpu/transforms/gemm_rewriter.cc

```diff
@@ -376,13 +378,52 @@ HloInstruction *TransposeMatrix(HloInstruction *instr, int64_t contracting_dim,
       non_contracting_dim = i;
     }
   }
-  permutation[non_contracting_dim] = contracting_dim;
-  permutation[contracting_dim] = non_contracting_dim;
+  if (!col_maj) {
+    permutation[non_contracting_dim] = contracting_dim;
+    permutation[contracting_dim] = non_contracting_dim;
 
-  Shape new_shape = ShapeUtil::PermuteDimensions(permutation, instr->shape());
-  *new_shape.mutable_layout() = instr->shape().layout();
-  return instr->AddInstruction(
-      HloInstruction::CreateTranspose(new_shape, instr, permutation));
+    Shape new_shape = ShapeUtil::PermuteDimensions(permutation, instr->shape());
+    *new_shape.mutable_layout() = instr->shape().layout();
+
+    return instr->AddInstruction(
+        HloInstruction::CreateTranspose(new_shape, instr, permutation));
+  }
+
+  Shape normalized_input_shape =
+      ShapeUtil::MakeShapeWithDescendingLayoutAndSamePhysicalLayout(
+          instr->shape());
+  auto a0 = MakeBitcastHlo(instr, normalized_input_shape);
+
+  int new_contracting_dim = -1;
+  int new_non_contracting_dim = -1;
+  for (int i = 0; i < instr->shape().dimensions_size(); ++i) {
+    auto dim = LayoutUtil::Major(instr->shape().layout(), i);
+    if (dim == contracting_dim) {
+      new_contracting_dim = i;
+    } else if (dim == non_contracting_dim) {
+      new_non_contracting_dim = i;
+    } else {
+      // Discard the batch dimensions.
+      permutation[i] = i;
+    }
+  }
+
+  permutation[new_non_contracting_dim] = new_contracting_dim;
+  permutation[new_contracting_dim] = new_non_contracting_dim;
+
+  Shape transpose_shape =
```

Comment:
I think computing permutation can be simplified quite a bit. You can compute `inv_perm` first (like below). Then

```
new_contracting_dim = inv_perm[contracting_dim];
new_non_contracting_dim = inv_perm[non_contracting_dim];
permutation = absl::c_iota(permutation, 0);
std::swap(permutation[new_contracting_dim], permutation[new_non_contracting_dim]);
```